twilight-gateway/src/shard.rs

```diff
@@ -669,455 +514,95 @@ impl<Q> Shard<Q> {
-    /// Returns a [`ReceiveMessageErrorType::FatallyClosed`] error type if the
-    /// shard was closed due to a fatal error, such as invalid authorization.
-    ///
-    /// Returns a [`ReceiveMessageErrorType::Process`] error type if the shard
-    /// failed to internally process a received event.
-    ///
-    /// Returns a [`ReceiveMessageErrorType::Reconnect`] error type if the shard
-    /// failed to reconnect to the gateway. This isn't a fatal error and can be
-    /// retried.
-    ///
-    /// Returns a [`ReceiveMessageErrorType::SendingMessage`] error type if the
-    /// shard failed to send a message to the gateway, such as a heartbeat.
-    #[tracing::instrument(fields(id = %self.id), name = "shard", skip(self))]
-    pub async fn next_message(&mut self) -> Result<Message, ReceiveMessageError> {
-        /// Actions the shard might take.
-        enum Action {
-            /// Close the gateway connection with this close frame.
-            Close(CloseFrame<'static>),
-            /// Send this command to the gateway.
-            Command(String),
-            /// Send a heartbeat command to the gateway.
-            Heartbeat,
-            /// Identify with the gateway.
-            Identify,
-            /// Handle this incoming gateway message.
-            Message(Option<Result<WebsocketMessage, WebsocketError>>),
-        }
 
-        match self.status {
-            ConnectionStatus::Disconnected {
-                close_code,
-                reconnect_attempts,
-            } => {
-                // The shard should should only reconnect after the gateway
-                // closes the underlying TCP connection.
-                if self.connection.is_none() {
-                    self.reconnect(close_code, reconnect_attempts).await?;
+    /// Send and flush the pending message.
+    fn poll_flush_pending(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), ReceiveMessageError>> {
+        ready!(Pin::new(self.connection.as_mut().unwrap())
+            .poll_ready(cx)
+            .map_err(|source| {
+                self.disconnect(CloseInitiator::Transport);
+                self.connection = None;
+                ReceiveMessageError::from_websocket(source)
+            }))?;
+
+        let Some(pending) = &mut self.pending else {
+            return Poll::Ready(Ok(()));
+        };
+        if let Some(item) = &pending.gateway_event {
+            if let Some(ratelimiter) = self.ratelimiter.as_mut() {
+                if item.is_text() && !pending.is_heartbeat {
+                    ready!(ratelimiter.poll_acquire(cx));
                 }
             }
-            ConnectionStatus::FatallyClosed { close_code } if self.connection.is_none() => {
-                return Err(ReceiveMessageError::from_fatally_closed(close_code));
+            if let Err(source) = Pin::new(self.connection.as_mut().unwrap())
+                .start_send(pending.gateway_event.take().unwrap().into_websocket_msg())
```

Comment:
I think this could use some comments clarifying that `take()` ensures we don't call `start_send` twice. (Overall there could be more comments to explain the cancellation safety, I'm scratching my head on some of these until I see what ensures it)